From 28d5940e328685dde29195f124bad70aa3d9c36f Mon Sep 17 00:00:00 2001
From: Martin Jansa <Martin.Jansa@gmail.com>
Date: Sat, 13 Mar 2021 18:36:16 +0100
Subject: [PATCH] rust-llvm-1.37.0: fix build with gcc-10 and undo incompatible
 changes from meta-rust

* drop RISCV from rust-llvm as it's only experimental in this version

* drop changelog-seen from rust.inc as it was added with 1.49.0 version, but is not supported in 1.37.0
    Finished dev [unoptimized] target(s) in 25.17s
running: /OE/build/test-oe-build-time/poky/build/tmp/work/x86_64-linux/rust-native/1.37.0-r0/rustc-1.37.0-src/build/bootstrap/debug/bootstrap -j 64 build --verbose
failed to parse TOML configuration 'config.toml': unknown field `changelog-seen`, expected one of `build`, `install`, `llvm`, `rust`, `target`, `dist`
Traceback (most recent call last):
  File "/OE/build/test-oe-build-time/poky/build/tmp/work/x86_64-linux/rust-native/1.37.0-r0/rustc-1.37.0-src/src/bootstrap/bootstrap.py", line 864, in <module>
    main()
  File "/OE/build/test-oe-build-time/poky/build/tmp/work/x86_64-linux/rust-native/1.37.0-r0/rustc-1.37.0-src/src/bootstrap/bootstrap.py", line 847, in main
    bootstrap(help_triggered)
  File "/OE/build/test-oe-build-time/poky/build/tmp/work/x86_64-linux/rust-native/1.37.0-r0/rustc-1.37.0-src/src/bootstrap/bootstrap.py", line 833, in bootstrap
    run(args, env=env, verbose=build.verbose)
  File "/OE/build/test-oe-build-time/poky/build/tmp/work/x86_64-linux/rust-native/1.37.0-r0/rustc-1.37.0-src/src/bootstrap/bootstrap.py", line 141, in run
    raise RuntimeError(err)
RuntimeError: failed to run: /OE/build/test-oe-build-time/poky/build/tmp/work/x86_64-linux/rust-native/1.37.0-r0/rustc-1.37.0-src/build/bootstrap/debug/bootstrap -j 64 build --verbose

* similarly drop embed-bitcode as it's also not compatible with 1.37.0
  | Caused by:
  |   process didn't exit successfully: `rustc - --crate-name ___ --print=file-names -L /OE/build/test-oe-build-time/poky/build/tmp/work/core2-64-poky-linux/libstd-rs/1.37.0-r0/recipe-sysroot/usr/lib/rust --remap-path-prefix=/OE/build/test-oe-build-time/poky/build/tmp/work/core2-64-poky-linux/libstd-rs/1.37.0-r0=/usr/src/debug/libstd-rs/1.37.0-r0 -Cembed-bitcode=yes -L /OE/build/test-oe-build-time/poky/build/tmp/work/core2-64-poky-linux/libstd-rs/1.37.0-r0/recipe-sysroot/usr/lib -C link-arg=-Wl,-soname,libstd.so --target x86_64-poky-linux --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro` (exit code: 1)
  | --- stderr
  | error: unknown codegen option: `embed-bitcode`

Signed-off-by: Martin Jansa <Martin.Jansa@gmail.com>
---
 recipes-devtools/rust/libstd-rs.inc           |  2 -
 ...752888c0282bf5504b67484b92d8d069f1b8.patch | 38 +++++++++++++++++++
 recipes-devtools/rust/rust-llvm_1.37.0.bb     |  8 ++++
 recipes-devtools/rust/rust.inc                |  1 -
 4 files changed, 46 insertions(+), 3 deletions(-)
 create mode 100644 recipes-devtools/rust/rust-llvm/98b1752888c0282bf5504b67484b92d8d069f1b8.patch

diff --git a/recipes-devtools/rust/libstd-rs.inc b/recipes-devtools/rust/libstd-rs.inc
index 1bf2dab..071ebaa 100644
--- a/recipes-devtools/rust/libstd-rs.inc
+++ b/recipes-devtools/rust/libstd-rs.inc
@@ -12,8 +12,6 @@ DEPENDS_append_libc-musl = " libunwind"
 DEPENDS_remove_riscv32 = "libunwind"
 DEPENDS_remove_riscv64 = "libunwind"
 
-# Embed bitcode in order to allow compiling both with and without LTO
-RUSTFLAGS += "-Cembed-bitcode=yes"
 # Needed so cargo can find libbacktrace
 RUSTFLAGS += "-L ${STAGING_LIBDIR} -C link-arg=-Wl,-soname,libstd.so"
 
diff --git a/recipes-devtools/rust/rust-llvm/98b1752888c0282bf5504b67484b92d8d069f1b8.patch b/recipes-devtools/rust/rust-llvm/98b1752888c0282bf5504b67484b92d8d069f1b8.patch
new file mode 100644
index 0000000..430205d
--- /dev/null
+++ b/recipes-devtools/rust/rust-llvm/98b1752888c0282bf5504b67484b92d8d069f1b8.patch
@@ -0,0 +1,38 @@
+From 98b1752888c0282bf5504b67484b92d8d069f1b8 Mon Sep 17 00:00:00 2001
+From: Than McIntosh <thanm@google.com>
+Date: Fri, 19 Jul 2019 13:13:54 +0000
+Subject: [PATCH] Merging r366572:
+
+------------------------------------------------------------------------
+r366572 | thanm | 2019-07-19 06:13:54 -0700 (Fri, 19 Jul 2019) | 12 lines
+
+[NFC] include cstdint/string prior to using uint8_t/string
+
+Summary: include proper header prior to use of uint8_t typedef
+and std::string.
+
+Subscribers: llvm-commits
+
+Reviewers: cherry
+
+Tags: #llvm
+
+Differential Revision: https://reviews.llvm.org/D64937
+------------------------------------------------------------------------
+---
+ include/llvm/Demangle/MicrosoftDemangleNodes.h | 2 ++
+ 1 file changed, 2 insertions(+)
+
+diff --git a/include/llvm/Demangle/MicrosoftDemangleNodes.h b/include/llvm/Demangle/MicrosoftDemangleNodes.h
+index da9d9d5bfdc0..3d47471f0ef0 100644
+--- a/include/llvm/Demangle/MicrosoftDemangleNodes.h
++++ b/include/llvm/Demangle/MicrosoftDemangleNodes.h
+@@ -16,6 +16,8 @@
+ #include "llvm/Demangle/DemangleConfig.h"
+ #include "llvm/Demangle/StringView.h"
+ #include <array>
++#include <cstdint>
++#include <string>
+ 
+ namespace llvm {
+ namespace itanium_demangle {
diff --git a/recipes-devtools/rust/rust-llvm_1.37.0.bb b/recipes-devtools/rust/rust-llvm_1.37.0.bb
index a40201c..3f18b89 100644
--- a/recipes-devtools/rust/rust-llvm_1.37.0.bb
+++ b/recipes-devtools/rust/rust-llvm_1.37.0.bb
@@ -1,2 +1,10 @@
 require rust-source-${PV}.inc
 require rust-llvm-ncsa.inc
+
+# Fix build with gcc-10
+SRC_URI += "file://98b1752888c0282bf5504b67484b92d8d069f1b8.patch"
+
+# remove RISCV, because with 1.37.0 version it needs to be passed
+# in LLVM_EXPERIMENTAL_TARGETS_TO_BUILD, not LLVM_TARGETS_TO_BUILD
+# EXTRA_OECMAKE_remove = ";RISCV"
+EXTRA_OECMAKE_append = " -DLLVM_TARGETS_TO_BUILD='ARM;AArch64;Mips;PowerPC;X86'"
diff --git a/recipes-devtools/rust/rust.inc b/recipes-devtools/rust/rust.inc
index 9dd9979..81fcb01 100644
--- a/recipes-devtools/rust/rust.inc
+++ b/recipes-devtools/rust/rust.inc
@@ -474,7 +474,6 @@ python do_configure() {
     config.set("install", "mandir",  e(d.getVar("D", True) + d.getVar("mandir", True)))
 
     with open("config.toml", "w") as f:
-        f.write('changelog-seen = 2\n\n')
         config.write(f)
 
     # set up ${WORKDIR}/cargo_home
-- 
2.30.2

